src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,194 @@
+                    head.next = tail;
+                } else {
+                    current = new Node<>(head, value, node);
+                    node.prev = current;
+                    head.next = current;
+                }
+                break;
+            default:
+                if (index == size) {
+                    current = new Node<>(tail, value, null);
+                    tail.next = current;
+                    tail = current;
+                } else {
+                    current = new Node<>(node.prev, value, node);
+                    node.prev.next = current;
+                    node.prev = current;
+                }
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        Node<T> node;
+        for (T value : list) {
+            this.add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIfIndexValid(index);
+        Node<T> node = getNode(index);
+        return node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIfIndexValid(index);
+        Node<T> node = getNode(index);
+        T current = node.value;
+        node.value = value;
+        return current;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIfIndexValid(index);
+        Node<T> node = getNode(index);
+        T value = node.value;
+        unlink(node);
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
+        if (size == 0) {
+            return false;
+        }
+        Node<T> node = head;
+        while (node != null) {
+            if (node.value == null && object == null
+                    || (node.value != null && node.value.equals(object))) {
+                unlink(node);
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() == 0;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node == head) {
+            switch (size) {
+                case 1:
+                    head = null;
+                    break;
+                case 2:
+                    head = tail;
+                    tail = null;
+                    head.prev = null;
+                    break;
+                default:
+                    head = node.next;
+                    head.prev = null;
+            }
+        } else if (node.next == null && size > 2) {
+            tail = node.prev;
+            tail.next = null;
+        } else if (node.next == null && size == 2) {
+            head.next = null;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
```

Comment:
```suggestion
        Node<T> prev = node.prev;
        Node<T> next = node.next;

        if (prev == null) {
            head = next;
        } else {
            prev.next = next;
            node.prev = null;
        }

        if (next == null) {
            tail = prev;
        } else {
            next.prev = prev;
            node.next = null;
        }

        node.value = null;
```
можна спростити це таким чином

Reply:
looks good